CoreFoundation/Base.subproj/CFKnownLocations.c

```diff
@@ -91,6 +91,20 @@ CFURLRef _Nullable _CFKnownLocationCreatePreferencesURLForUser(CFKnownLocationUs
             break;
     }
 
+#elif TARGET_OS_ANDROID
+
+    switch (user) {
+    case _kCFKnownLocationUserAny:
+    case _kCFKnownLocationUserByName:
+      abort();
+    case _kCFKnownLocationUserCurrent: {
+      CFURLRef home = CFCopyHomeDirectoryURLForUser(CFGetUserName());
+      location = CFURLCreateWithFileSystemPathRelativeToBase(kCFAllocatorSystemDefault, CFSTR("/Apple/Library/Preferences"), kCFURLPOSIXPathStyle, true, home);
+      CFRelease(home);
+      break;
```

Comment:
I'm not sure this is going to work on the device. The home directory in Android is actually the root directory, and it is not writable by the user, so you will not be able to create any new directories there. Have you tried this on-device?

Reply:
No, I don't have a dvice on hand right now.  The desire here is to get the path to the application's data space.  What would be the way to get that?

Reply:
There's no such thing in Android, sadly. Android (Java) has the `SharedPreferences` type, but the physical storage is not public. If you want something that will work, you can use the temporary directory, which is readable/writable by any process. I really don’t think you can give a value here that makes sense in Android.

Reply:
Can we assume that we have the Java runtime initialized and can just JNI to get the value?  Using tmp seems .... less than ideal.  This is expected to be used for preserved data.

Reply:
You still have the problem that you want to return a path, while `SharedPrefereces` is a database.

One solution use in the HTTP support is checking an environment variable to pick up the certificates from an app known location. Maybe this needs a similar solution for Android. Only the app can decide which file system access they are going to have, and where it is going to happen. Even if you have access to JNI, for many APIs you need a reference to `Context`, which will be difficult to obtain.

```java
// A some moment during the start of your Android app
Context context = … // this is the piece that will be quite difficult to obtain in CF
File filesDir = context.getFilesDir();
android.system.Os.setenv("CFAndroidFileDirPath", filesDir.getAbsolutePath(), true);
```

```c
// In this file, or somewhere in CF
const char *filesDirPtr = getenv("CFAndroidFileDirPath");
CFStringRef filesDirPath = CFStringCreateWithCString(NULL, filesDirPtr, kCFStringEncodingUTF16); // NOTE: I think Java uses UTF-16
CFURLRef filesDirURL = CFURLCreateWithFileSystemPAth(NULL, filesDirPath, kCFURLPOSIXPathStyle, true);
location = CFURLCreateWithFileSystemPathRelativeToBase(kCFAllocatorSystemDefault, CFSTR("/Library/Preferences"), kCFURLPOSIXPathStyle, true, filesDirURL);
```

Reply:
Sure, I think that using an environment variable is plausible, but, what would the developer set that to?  This is the *common* case - I just want persistent storage of files